Fix: bad operator precedence inside object

jods4 opened the pull request:
This filter:
```js
{ 
  a: 1, 
  or: { b: 2, c: 3 },
}
```
was incorrectly converted to: `a eq 1 and b eq 2 or c eq 3`.

Brackets are required to fix the precedence of `or` as expressed in the object structure.

After this fix:
`a eq 1 and (b eq 2 or c eq 3)`

techniq commented:
@jods4 Thanks for the PR.  Can you take a look at the 2 breaking tests after this change, and add a new test for this use case (if there isn't one covering it already)?  Thanks.

jods4 commented:
Hi @techniq.

I just checked the failing tests, they are purely based on textual expectations.
There is no behavioural regression, the test queries all have the exact same semantics, but some tests fail because of extra brackets where there was none before; admittedly in less than ideal places.

I think they're worth discussing before I attempt to "fix" them.

### It should handle simple logical operators (and, or, etc) as an object (no parens)
```js
const filter = { and: { SomeProp: 1, AnotherProp: 2 } };
```
```
Expected: "?$filter=SomeProp eq 1 and AnotherProp eq 2"
Received: "?$filter=(SomeProp eq 1 and AnotherProp eq 2)"
```
I think this test is the root cause of the bug. 
The "no parens" here clearly hints at the fact that it's intentional, but when you nest the test case as:
```js
const filter = { Whatever: 0, or: { SomeProp: 1, AnotherPro: 2 } };
```
that lack of parens becomes a bug.

### It should handle nested logical operators
```js
const filter = {
  and: [{ SomeProp: 1 }, { or: [{ AnotherProp: 2 }, { ThirdProp: 3 }] }],
};
```
```
Expected: "?$filter=((Prop1 eq 1) or (Prop2 eq 2 and Prop3 eq 3))"
Received: "?$filter=((Prop1 eq 1) or ((Prop2 eq 2 and Prop3 eq 3)))"
```
Here the query nesting has generated doubled parens around the second `or` operand.
Notice there is an unneeded parens pair around the whole expression, much like in previous failure, that was already there before.

### Handling brackets properly
I fixed the bug simply by adding extra hard-coded brackets where they might be needed, as was already done in other places.

Generating clean odata queries with brackets only where required is a really nice goal, but it's hard (impossible?) to achieve without tracking an expression tree and knowing operators precedence at each node. I don't think it's achievable without significant changes in the library.

Maybe it could be possible to optimize brackets generation a bit more, while erring on the side of caution (correctness is more important) but that's tricky.

The easy "fix" here is to change expectations on these two tests and go with the extra, but correct, brackets.

What do you think?

techniq commented:
@jods4 I agree correctness is more important than optimal readability (and it's not a huge reduction).  If you could update the failing tests accordingly, I'll merge and push a release in the next day or two.

jods4 commented:
@techniq ok I've changed the expectations for the 2 failing tests, and I added 1 new test that checks the bug I reported and fixed in this PR.

No hurry for the release, I'm using my own fork of your project anyway.
I just figured I would contribute back some of the changes I made.
